nym-vpn-core/crates/nym-gateway-probe/src/lib.rs

```diff
@@ -77,20 +96,143 @@ pub async fn probe(entry_point: EntryPoint) -> anyhow::Result<ProbeResult> {
-        outcome,
+    // Disconnect the mixnet client gracefully
+    outcome.map(|mut outcome| {
+        outcome.wg = Some(wg_outcome);
+        ProbeResult {
+            gateway: entry_gateway.clone(),
+            outcome,
+        }
     })
 }
 
+async fn wg_probe(
+    authenticator: AuthAddress,
+    shared_mixnet_client: Arc<Mutex<Option<MixnetClient>>>,
+    gateway_host: &str,
+) -> anyhow::Result<Wg> {
+    let auth_shared_client =
+        nym_authenticator_client::SharedMixnetClient::from_shared(&shared_mixnet_client);
+    let mut auth_client = nym_authenticator_client::AuthClient::new(auth_shared_client).await;
+
+    let mut rng = rand::thread_rng();
+    let private_key = nym_crypto::asymmetric::encryption::PrivateKey::new(&mut rng);
+    let public_key = private_key.public_key();
+
+    let init_message = ClientMessage::Initial(InitMessage {
+        pub_key: PeerPublicKey::new(public_key.to_bytes().into()),
+    });
+
+    let mut wg_outcome = Wg {
+        can_register: false,
+        can_handshake: false,
+        can_resolve_dns: false,
+        ping_hosts_performance: 0.,
+        ping_ips_performance: 0.,
+    };
+
+    if let Some(authenticator_address) = authenticator.0 {
+        let response = auth_client
+            .send(init_message, authenticator_address)
+            .await?;
+
+        let registered_data = match response.data {
+            AuthenticatorResponseData::PendingRegistration(PendingRegistrationResponse {
+                reply:
+                    RegistrationData {
+                        nonce,
+                        gateway_data,
+                        ..
+                    },
+                ..
+            }) => {
+                // Unwrap since we have already checked that we have the keypair.
+                debug!("Verifying data");
+                gateway_data.verify(&private_key, nonce)?;
+
+                let finalized_message = ClientMessage::Final(GatewayClient::new(
+                    &private_key,
+                    gateway_data.pub_key().inner(),
+                    gateway_data.private_ip,
+                    nonce,
+                ));
+                let response = auth_client
+                    .send(finalized_message, authenticator_address)
+                    .await?;
+                let AuthenticatorResponseData::Registered(RegisteredResponse { reply, .. }) =
+                    response.data
+                else {
+                    bail!("Unexpected response: {response:?}");
+                };
+                reply
+            }
+            AuthenticatorResponseData::Registered(RegisteredResponse { reply, .. }) => reply,
+            _ => bail!("Unexpected response: {response:?}"),
+        };
+
+        let peer_public = registered_data.pub_key.inner();
+        let static_private = x25519_dalek::StaticSecret::from(private_key.to_bytes());
+
+        // let private_key_bs64 = general_purpose::STANDARD.encode(static_private.as_bytes());
```

Comment:
forgotten comment?

Reply:
No, that one is actually useful for debugging, no harm leaving it in there